gadopt/gplates.py

```diff
@@ -0,0 +1,444 @@
+
+    Methods:
+        update_plate_reconstruction(ndtime):
+            Updates the function values based on plate velocities from GPlates
+            for a given model time.
+            **Note** model time is non-dimensionalised
+
+    Example:
+        >>> # Assuming necessary imports and setup are done
+        >>> gplates_function = GplatesFunction(V,
+        ...                                    gplates_connector=pl_rec_model,
+        ...                                    name="GplateVelocity")
+        >>> gplates_function.update_plate_reconstruction(ndtime=0.0)
+    """
+    def __new__(cls, *args, **kwargs):
+        # Ensure compatibility with Firedrake Function's __new__ method
+        return super().__new__(cls, *args, **kwargs)
+
+    def __init__(self, function_space, *args, gplates_connector=None, top_boundary_marker="top", **kwargs):
+        # Initialize as a Firedrake Function
+        super().__init__(function_space, *args, **kwargs)
+
+        # Cache all the necessary information that will be used to assign surface velocities
+        # the marker for surface boundary. This is typically "top" in extruded mesh.
+        self.top_boundary_marker = top_boundary_marker
+        # establishing the DirichletBC that will be used to find surface nodes
+        self.dbc = fd.DirichletBC(
+            self.function_space(),
+            self,
+            sub_domain=self.top_boundary_marker)
+        # coordinates of surface points
+        self.boundary_coords = fd.Function(
+            self.function_space(),
+            name="coordinates").interpolate(
+                fd.SpatialCoordinate(
+                    extract_unique_domain(self))).dat.data_ro_with_halos[self.dbc.nodes]
+        self.boundary_coords /= np.linalg.norm(self.boundary_coords, axis=1)[:, np.newaxis]
+
+        # Store the GPlates connector
+        self.gplates_connector = gplates_connector
+
+
+class pyGplatesConnector(object):
+    # Non-dimensionalisation constants
+    # Factor to non-dimensionalise gplates velocities: d/kappa
+    #   u [metre/sec] * velocity_non_dim_factor = u []
+    velocity_non_dim_factor = 2890e3 / 1.0e-6
+    # Factor to dimensionalise model time: d^2/kappa
+    #   t [] X time_dim_factor = t [sec]
+    time_dim_factor = 2890e3**2 / 1.0e-6
+    #   t [year] X yrs2sec = t [sec]
+    yrs2sec = 365 * 24 * 60 * 60
+    #   t [Myrs] X myrs2sec = t[sec]
+    myrs2sec = 1e6*yrs2sec
+    #   t [] X time_dimDmyrs2sec = t [Myrs]
+    time_dimDmyrs2sec = time_dim_factor / myrs2sec
+    #   L [cm] X cm2m = L [m]
+    cm2m = 1e-2
+    #   u [cm/year] * velocity_dimDcmyr = u []
+    velocity_dimDcmyr = velocity_non_dim_factor * cm2m / yrs2sec
+
+    # minimum distance, bellow which we do not interpolate
+    #   this is just to avoid division by zero when weighted averaging
+    epsilon_distance = 1e-8
+
+    def __init__(self, rotation_filenames, topology_filenames, oldest_age, delta_t=1., scaling_factor=1.0, nseeds=1000, nneighbours=3):
```

Comment:
Are you happy with your default parameter choice for nseeds and nneighbours here? When we tested this together around a month ago, we found this made a difference. For our previous GPlates simulations (in the 2022 paper and Tom's stuff), we have used  nseeds=1e5, nneighbours=4. This seems to lead to sharper velocity gradients at the plate boundaries. Some may say this is a better representation of rigid plate motions, but it also makes the Stokes solution process more challenging, so there is a balance here. I'm happy with these defaults if you are, but perhaps add a comment similar to what I've written here, so that we remember this discussion!

Reply:
Done